Userland/Libraries/LibWeb/CSS/StyleComputer.cpp

```diff
@@ -953,6 +966,319 @@ static ErrorOr<void> cascade_custom_properties(DOM::Element& element, Optional<C
+        valid_properties++;
+    }
+
+    dbgln_if(LIBWEB_CSS_ANIMATION_DEBUG, "Animation {} contains {} properties to interpolate, progress = {}%", name, valid_properties, progress_in_keyframe * 100);
+
+    if (fill_mode == Forward || fill_mode == Both) {
+        if (!active_state_if_fill_forward)
+            active_state_if_fill_forward = make<AnimationStateSnapshot>();
+    }
+
+    UnderlyingType<PropertyID> property_id_value = 0;
+    for (auto const& property : keyframe_values.resolved_properties) {
+        auto property_id = static_cast<PropertyID>(property_id_value++);
+        if (property.has<Empty>())
+            continue;
+
+        auto resolve_property = [&](auto& property) {
+            return property.visit(
+                [](Empty) -> RefPtr<StyleValue const> { VERIFY_NOT_REACHED(); },
+                [&](AnimationKeyFrameSet::ResolvedKeyFrame::UseInitial) {
+                    if (auto value = initial_state.state[to_underlying(property_id)])
+                        return value;
+
+                    auto value = style_properties.maybe_null_property(property_id);
+                    initial_state.state[to_underlying(property_id)] = value;
+                    return value;
+                },
+                [&](RefPtr<StyleValue const> value) { return value; });
+        };
+
+        auto resolved_start_property = resolve_property(property);
+
+        auto const& end_property = keyframe_end_values.resolved_properties[to_underlying(property_id)];
+        if (end_property.has<Empty>()) {
+            if (resolved_start_property) {
+                style_properties.set_property(property_id, resolved_start_property.release_nonnull());
+                dbgln_if(LIBWEB_CSS_ANIMATION_DEBUG, "No end property for property {}, using {}", string_from_property_id(property_id), resolved_start_property->to_string());
+            }
+            continue;
+        }
+
+        auto resolved_end_property = resolve_property(end_property);
+
+        if (!resolved_start_property || !resolved_end_property)
+            continue;
+
+        auto start = resolved_start_property.release_nonnull();
+        auto end = resolved_end_property.release_nonnull();
+
+        // FIXME: This should be a function of the animation-timing-function.
+        auto next_value = TRY(interpolate_property(*start, *end, progress_in_keyframe));
+        dbgln_if(LIBWEB_CSS_ANIMATION_DEBUG, "Interpolated value for property {} at {}: {} -> {} = {}", string_from_property_id(property_id), progress_in_keyframe, start->to_string(), end->to_string(), next_value->to_string());
+        style_properties.set_property(property_id, next_value);
+        if (active_state_if_fill_forward)
+            active_state_if_fill_forward->state[to_underlying(property_id)] = next_value;
+    }
+
+    return {};
+}
+
+bool StyleComputer::Animation::is_done() const
+{
+    return progress.as_fraction() >= 0.9999f && iteration_count.has_value() && iteration_count.value() == 0;
```

Comment:
there was a different place with three 9s previously — looks like a constant we could unify…? (some kind of `_threshold` or `_epsilon` as people tend to name it?)

Reply:
Unless it shows up as a broken animation, the actual number here doesn't really matter (you'd only observe this on say, `animation-duration: 10000s;` where 99.99% is 1 second, but even then, that error is fine imo for an initial impl, the "real" spec-compliant implementation of this function has other parameters anyway)